### PR TITLE
[parsing] Add SDFormat support for //world/joint

### PR DESCRIPTION
**Pull request by liamhan0905**
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322)

### CODE REVIEW

File: multibody/parsing/detail_sdf_parser.cc [hunk 1]
```diff
@@ -1755,6 +1766,18 @@ std::vector<ModelInstanceIndex> AddModelsFromSdf(
...
           workspace.plant->world_frame(), workspace.plant);
     }
+
+    // Add all the joints
+    std::set<sdf::JointType> joint_types;
+    for (uint64_t joint_index = 0; joint_index < world.JointCount();
+        ++joint_index) {
+      const sdf::Joint& joint = *world.JointByIndex(joint_index);
+      // const sdf::Model& model = *world.ModelByIndex(joint_index);
+      // const RigidTransformd X_WM;
```
Comment (liamhan0905) [inline multibody/parsing/detail_sdf_parser.cc:1776]: I'll get rid of these on the next push

**Review by azeey — COMMENTED** [review]
Looks good so far. My main concern is that we may not be taking the joint pose into account.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NGsnEgR5YZgSTmQERQG:b-oakghp)* status: 6 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @liamhan0905)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 111 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsl46w3DMGNb9Qjnm1:-NGsl46w3DMGNb9Qjnm2:b-214nlc) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L111)):*
> ```cpp
> // a child of a nested model, the local name of the body is prefixed with the
> // scoped name of the nested model. Furthermore, if the body is a child of a
> // model encapsulated inside the world, the local name of the body is prefixed
> ```

This addition only applies when the `relative_to_model_instance` is the `world_model_instance` right? Otherwise, it would return the local name without a prefix as the first part of the comment paragraph suggests.
___
*[`multibody/parsing/detail_sdf_parser.cc` line 121 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsko5-5HKvJFXCk5in:-NGsko5-5HKvJFXCk5io:b-2w6rjd) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L121)):*
> ```cpp
>   // If the body is a child of a model encapsulated inside the world, we need
>   // to prefix the body name with the model name
>   if(relative_to_model_absolute_name == "WorldModelInstance"){
> ```

nit: Probably better to compare `relative_to_model_instance` to `world_model_instance()` than a string comparison.
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1778 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGshh2D8_SSWJpOh176:-NGshh2D8_SSWJpOh177:bmyfi3t) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L1778)):*
> ```cpp
>       // const RigidTransformd X_WM;
>       AddJointFromSpecification(
>           workspace.diagnostic, {}, {}, joint, world_model_instance(),
> ```

I don't think we'd want to pass a `{}` for a `const sdf::Model &`. From what I can tell, the only place `model_spec`is used in `AddJointFromSpecification`, it is passed to `ExtractJointAxis`, which explicitly states that the argument is unused https://github.com/RobotLocomotion/drake/blob/15e90cbda7aeb2497ea32144e743b668043f3d6e/multibody/parsing/detail_sdf_parser.cc#L313. So my suggestion would be to remove the `sdf::Model` parameter entirely from `ExtractJointAxis` and `AddJointFromSepcification`
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1778 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGskJ5E9IKYs8u3puLX:-NGskJ5E9IKYs8u3puLY:b-2yd0kf) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L1778)):*
> ```cpp
>       // const RigidTransformd X_WM;
>       AddJointFromSpecification(
>           workspace.diagnostic, {}, {}, joint, world_model_instance(),
> ```

I think we'll need to test with the joint parent set to `world` to see the effect of passing `X_WM = {}` here. 

In 
https://github.com/RobotLocomotion/drake/blob/15e90cbda7aeb2497ea32144e743b668043f3d6e/multibody/parsing/detail_sdf_parser.cc#L635-L638

the joint pose is being resolved relative to `__model__`, but if the joint is in the world, I don't think that call would succeed.
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 2596 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsm03P4Ux8Ow3UvQv7:-NGsm03P4Ux8Ow3UvQv8:blw7dum) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/test/detail_sdf_parser_test.cc#L2596)):*
> ```cpp
>   plant_.Finalize();
> 
>   ASSERT_TRUE(plant_.HasModelInstanceNamed("WorldModelInstance"));
> ```

nit: We can probably skip this check since it's a Drake invariant. It would have been tested elsewhere.
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 2626 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsmpar9d2fsxGAytMU:-NGsmpar9d2fsxGAytMV:b6c9u23) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/test/detail_sdf_parser_test.cc#L2626)):*
> ```cpp
>   EXPECT_TRUE(joint.parent_body().name() == "L_P");
>   EXPECT_TRUE(joint.child_body().name() == "L_C");
> ```

To check that the `<pose>` tag inside `<joint>` is working, can you get the joint's `frame_on_parent` and `frame_on_child` and add expectations on their pose?

**Review by liamhan0905 — COMMENTED** [review]
I think that the joint pose will be addressed by the last comment you made? Or are you referring to a different thing?
```
To check that the <pose> tag inside <joint> is working, can you get the joint's frame_on_parent and frame_on_child and add expectations on their pose?
```

I left some questions but will modify the ones that I wrote as "sounds good to me". Thanks for taking your time to review!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NGtwGo12UonVBkEo6Bh:bql5r5s)* status: 4 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 111 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsl46w3DMGNb9Qjnm1:-NGtalya-bUijuFeMrcZ:b-jtym8p) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L111)):*
> This addition only applies when the relative_to_model_instance is the world_model_instance right? Otherwise, it would return the local name without a prefix as the first part of the comment paragraph suggests.

Yes, but to be precise it'll also check for nested model. Below is the logic 
```
if (check if `relative_to_model_instance` is world model instance)
     return model string + body string
else if (check if body model instance != `relative_to_model_instance`)
     return nested model string + body string
else
     return body name
```
Technically, the first if statement (checking the world instance) can be put inside the else if block and just have if and else statement instead of if, else if and else statement as shown below. This is valid since `relative_to_model_instance` will be the world which will never be the same as the body model instance. But I'm not a big fan of nested if statements and thought this might be cleaner. But if you think otherwise let me know :)

```
if (check if body model instance != `relative_to_model_instance`)
  // moved inside 
  if (check if `relative_to_model_instance` is world model instance)
     return model string + body string
  //   
return nested model string + body string
else
     return body name
```
___
*[`multibody/parsing/detail_sdf_parser.cc` line 121 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsko5-5HKvJFXCk5in:-NGt_yzJ3_hARi_9bboU:btahtyx) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L121)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

nit: Probably better to compare `relative_to_model_instance` to `world_model_instance()` than a string comparison.
</blockquote></details>

sounds good to me
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1778 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGshh2D8_SSWJpOh176:-NGtj8by9Xlt3vswvHUl:b-joddjj) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L1778)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

I don't think we'd want to pass a `{}` for a `const sdf::Model &`. From what I can tell, the only place `model_spec`is used in `AddJointFromSpecification`, it is passed to `ExtractJointAxis`, which explicitly states that the argument is unused https://github.com/RobotLocomotion/drake/blob/15e90cbda7aeb2497ea32144e743b668043f3d6e/multibody/parsing/detail_sdf_parser.cc#L313. So my suggestion would be to remove the `sdf::Model` parameter entirely from `ExtractJointAxis` and `AddJointFromSepcification`
</blockquote></details>

sounds good to me
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1778 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGskJ5E9IKYs8u3puLX:-NGtqkagD5TlvUQs48pi:bxlae4c) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L1778)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

I think we'll need to test with the joint parent set to `world` to see the effect of passing `X_WM = {}` here. 

In 
https://github.com/RobotLocomotion/drake/blob/15e90cbda7aeb2497ea32144e743b668043f3d6e/multibody/parsing/detail_sdf_parser.cc#L635-L638

the joint pose is being resolved relative to `__model__`, but if the joint is in the world, I don't think that call would succeed.
</blockquote></details>

If I set the parent of the joint to be the world, I do indeed get an error. Couple of follow up questions

- Will the `__model__` refer to the world frame since the parent of the joint is the world? Or can this keyword not be used since the joint is outside the `model` tag?
- I'm not following how you were able to deduce that this call wouldn't work. Could you walk me through?

```
    1 <?xml version="1.0" ?>
    2 <sdf version="1.10"> 
    3   <world name="world_joint_child_frame">
    4     <model name="parent_model">
    5       <pose>0 0 1 0 0 0</pose>
    6       <link name="L_P"/>
    7     </model>
    8     <model name="child_model">
    9       <pose>0 0 10 0 0 0</pose>
   10       <link name="L_C"/>
   11     </model>
   12     <frame name="child_frame" attached_to="child_model">
   13       <pose>0 1 0 0 0 0</pose>
   14     </frame>
   15     <joint name="J2" type="ball">
   16       <pose>1 0 0 0 0 0</pose>
   17       <parent>world</parent>
   18       <child>parent_model::L_P</child>
   19     </joint>
   20   </world>
   21 </sdf>

```
```
[ RUN      ] SdfParserTest.WorldJoint
bazel-out/k8-dbg/bin/multibody/parsing/_virtual_includes/diagnostic_policy_test_base/drake/multibody/parsing/test/diagnostic_policy_test_base.h:63: Failure
Value of: error_records_.empty()
  Actual: false
Expected: true
error: PoseRelativeToGraph unable to find unique frame with name [__model__] in graph.
```
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 2596 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsm03P4Ux8Ow3UvQv7:-NGtZPLCEueiVhNyk3hD:btahtyx) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/test/detail_sdf_parser_test.cc#L2596)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

nit: We can probably skip this check since it's a Drake invariant. It would have been tested elsewhere.
</blockquote></details>

sounds good to me
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 2626 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsmpar9d2fsxGAytMU:-NGtvvcw9tUiJv63bNzC:b-2bb2cn) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/test/detail_sdf_parser_test.cc#L2626)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

To check that the `<pose>` tag inside `<joint>` is working, can you get the joint's `frame_on_parent` and `frame_on_child` and add expectations on their pose?
</blockquote></details>

you mean something simliar to [this](https://github.com/RobotLocomotion/drake/blob/master/multibody/parsing/test/detail_sdf_parser_test.cc#L2253-L2267)?

**Comment by jwnimmer-tri** [comment]
It seems like we have +@azeey for feature review.  Let me know if I've misunderstood.

@drake-jenkins-bot ok to test

**Review by azeey — COMMENTED** [review]
Yes. Probably good if @EricCousineau-TRI would give it look after I'm done.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NGxT3t-7A3-fj4KauKa:b429pzx)* status: 4 unresolved discussions, LGTM missing from assignee azeey, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @liamhan0905)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 111 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsl46w3DMGNb9Qjnm1:-NGxQI9F0_72kayNYMDP:buv2vwj) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L111)):*
> Yes, but to be precise it'll also check for nested model. Below is the logic

I understand it checks for nested models, but that's already mentioned in the comment. But the addition in this PR 

"Furthermore, if the body is a child of a
// model encapsulated inside the world, the local name of the body is prefixed
// with the model name."

is a little misleading. It's *not* the fact that a body is a child of a model encapsulated inside the world that exercises the new logic, but that `relative_to_model_instance` is set to `world_model_instance`. 
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1778 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGskJ5E9IKYs8u3puLX:-NGxRbaaBERe87Qa6I9p:b-x2fwqy) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L1778)):*
`__model__` will not refer to the world frame. If you want to get the pose of the joint relative to `world`, you have to use `world` as the argument to `ResolveRigidTransform`.

> I'm not following how you were able to deduce that this call wouldn't work. Could you walk me through?

The code I linked to (starting from line 635) checks if the parent is `world` and tries to resolve the pose relative to `__model__`, which wouldn't work for world joints. We'll need some way of checking if the joint is a model joint or a world joint. Since you can't find this out from an `sdf::Joint`, the easiest thing to do is to add a parameter to `AddJointFromSpecification` that indicates whether the joint is a world or model joint.
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 2626 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsmpar9d2fsxGAytMU:-NGxSuyvAKM_lHFzc2FE:bb77ufg) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/test/detail_sdf_parser_test.cc#L2626)):*
<details><summary><i>Previously, liamhan0905 (Liam Han) wrote…</i></summary><blockquote>

you mean something simliar to [this](https://github.com/RobotLocomotion/drake/blob/master/multibody/parsing/test/detail_sdf_parser_test.cc#L2253-L2267)?
</blockquote></details>

Yes.

**Review by liamhan0905 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322)* status: 4 unresolved discussions, LGTM missing from assignee azeey, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 111 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsl46w3DMGNb9Qjnm1:-NGxXoxUCBob-JblUPh2:b7bp6ke) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L111)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

> Yes, but to be precise it'll also check for nested model. Below is the logic

I understand it checks for nested models, but that's already mentioned in the comment. But the addition in this PR 

"Furthermore, if the body is a child of a
// model encapsulated inside the world, the local name of the body is prefixed
// with the model name."

is a little misleading. It's *not* the fact that a body is a child of a model encapsulated inside the world that exercises the new logic, but that `relative_to_model_instance` is set to `world_model_instance`. 


</blockquote></details>

Ahh I see how that can be misleading. Looking back I'm not sure why I phrased it this way lol I'll go ahead and modify my comment to be more clear
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1778 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGskJ5E9IKYs8u3puLX:-NGxchXZ-_IofgPRbqc4:b-8asty5) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/detail_sdf_parser.cc#L1778)):*
> The code I linked to (starting from
Makes sense. I'll take your advice. Thanks!
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 2626 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NGsmpar9d2fsxGAytMU:-NGxWFGw9xZdIpk2uL8m:b-hfghm4) ([raw file](https://github.com/robotlocomotion/drake/blob/91b55add97a44b324e629fb8f12a223114c7115a/multibody/parsing/test/detail_sdf_parser_test.cc#L2626)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

Yes.
</blockquote></details>

Sounds good

**Review by EricCousineau-TRI — COMMENTED** [review]
Yup! +@ericcousineau-tri for platform pending Addisu's feature approval - thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NH4y_fZ87TlLsjhpapa:b-yb6bpd)* status: 4 unresolved discussions, LGTM missing from assignees azeey,EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey)

**Review by EricCousineau-TRI — COMMENTED** [review]
+@sammy-tri for additional platform review after Addisu and I do our rounds, double thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NH5FWHcAW1t5x0wbVpc:b-80siqj)* status: 4 unresolved discussions, LGTM missing from assignees azeey,EricCousineau-TRI(platform),sammy-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey and @sammy-tri)

**Comment by jwnimmer-tri** [comment]
For me at least, the PR title is not particularly illuminating.  Is there a more precise phrasing of the new feature that y'all could use as the title (... which will be the used as the commit subject during squash, and so will appear in the release notes).

**Review by liamhan0905 — COMMENTED** [review]
>  the new feature that y'all could use as the title (... which will be the used as the commit subject during squash, and so will appear in the release notes
@jwnimmer-tri how about this? "support adding joints directly under the <world> tag using `//world/joint`element ". @azeey if you can better articulate it, feel free to add/modify

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NH5GR4HEYsN1QK-xEmk:bhz7odp)* status: 4 unresolved discussions, LGTM missing from assignees sammy-tri(platform),azeey,EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey, @EricCousineau-TRI, and @sammy-tri)

**Comment by jwnimmer-tri** [comment]
In that case, maybe this:

 `[parsing] Add SDFormat support for //world/joint`

**Review by liamhan0905 — COMMENTED** [review]
> support adding joints directly under the tag
forgot the word "world". meant to say *support adding joints directly under the world tag using...
But i think what you have is more concise

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NH5KlGh-GTcE23jruuf:b-cuul8o)* status: 4 unresolved discussions, LGTM missing from assignees sammy-tri(platform),azeey,EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey, @EricCousineau-TRI, and @sammy-tri)

**Review by azeey — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322)* status: 3 unresolved discussions, LGTM missing from assignees sammy-tri(platform),azeey,EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @EricCousineau-TRI, @liamhan0905, and @sammy-tri)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 603 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5Xz0_3PN-Fgq1C3-j:-NH5Xz0_3PN-Fgq1C3-k:b-9q4liu) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L603)):*
> ```cpp
> // specification object.
> void AddJointFromSpecification(
>     const DiagnosticPolicy& diagnostic, const RigidTransformd& X_WM,
> ```

nit: Can you add a comment that `X_WM` should be identity if this function is called for adding world joints since a world joint doesn't have a containing model and `M = W` in that case.
___
*[`multibody/parsing/detail_sdf_parser.cc` line 641 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5TAz0DWEJ8sFLepDM:-NH5TAz0DWEJ8sFLepDN:b6kgmxc) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L641)):*
> ```cpp
>   // a "world" link from which we can request its pose (which in that case would
>   // be the identity).
>   std::string relative_to = (is_model_joint) ? "__model__" : "world";
> ```

nit: This can be const.
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1774 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5Uk4QAwDXbBIkkLhx:-NH5Uk4QAwDXbBIkkLhy:bxut1qh) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L1774)):*
> ```cpp
>       AddJointFromSpecification(
>           workspace.diagnostic, {}, joint, world_model_instance(),
>           workspace.plant, &joint_types, false);
> ```

@ericcousineau-tri Do we want to limit world joints to weld joints only?

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322)* status: 3 unresolved discussions, LGTM missing from assignees sammy-tri(platform),azeey,EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey, @liamhan0905, and @sammy-tri)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 1774 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5Uk4QAwDXbBIkkLhx:-NH5ljJE8GfZ1bdO7X8b:b-c5185u) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L1774)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

@ericcousineau-tri Do we want to limit world joints to weld joints only?
</blockquote></details>

Yup! Let's keep it simple, and expand only if there's a workflow that can't be achieved via in-model articulation and weld.

**Review by liamhan0905 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322)* status: LGTM missing from assignees sammy-tri(platform),azeey,EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @azeey and @sammy-tri)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 603 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5Xz0_3PN-Fgq1C3-j:-NH676nE5euE857Zc9pu:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L603)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

nit: Can you add a comment that `X_WM` should be identity if this function is called for adding world joints since a world joint doesn't have a containing model and `M = W` in that case.
</blockquote></details>

Done
___
*[`multibody/parsing/detail_sdf_parser.cc` line 641 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5TAz0DWEJ8sFLepDM:-NH7YfL97MfHXZlws5WA:b-896fjx) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L641)):*
<details><summary><i>Previously, azeey (Addisu Z. Taddese) wrote…</i></summary><blockquote>

nit: This can be const.
</blockquote></details>

Done
___
*[`multibody/parsing/detail_sdf_parser.cc` line 1774 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NH5Uk4QAwDXbBIkkLhx:-NH7XPjT3lO-49q9pQvc:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/710351ba15d3c6cb50d7664f22a27f659ab231b4/multibody/parsing/detail_sdf_parser.cc#L1774)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Yup! Let's keep it simple, and expand only if there's a workflow that can't be achieved via in-model articulation and weld.
</blockquote></details>

Done

**Review by azeey — COMMENTED** [review]
:lgtm: @ericcousineau-tri PTAL.

> Reviewed 3 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NHAwZufEqRc0ZHtwMac:b-7g33o1)* status: LGTM missing from assignees sammy-tri(platform),EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sammy-tri)

**Review by sammy-tri — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NI3kIoUAl3WcObuQl2V:bnfp4nl)* status: LGTM missing from assignee EricCousineau-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @liamhan0905)

**Review by EricCousineau-TRI — COMMENTED** [review]
:lgtm_strong:, thanks!

> Reviewed 1 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18322#-:-NIFvZap6vpXOA7JboYF:b5zo9cn)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @liamhan0905)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 1779 at r3](https://reviewable.io/reviews/robotlocomotion/drake/18322#-NIFuUIN8OS1RyDhmQD2:-NIFuUIN8OS1RyDhmQD3:b9hv50z) ([raw file](https://github.com/robotlocomotion/drake/blob/73ff52458d9f72c01287dedb2cceae887872bce3/multibody/parsing/detail_sdf_parser.cc#L1779)):*
> ```cpp
>     }
> 
>     for (sdf::JointType joint_type : joint_types) {
> ```

BTW Was going to ask if we could add this check before `AddJointFromSpecification`.
However, our parsing isn't necessarily transactional so it's kinda moot.
(i.e. if parsing fails, the plant/scene graph won't be restored to original state - users may do this in Python)

**Pull request merged**